util/unifyfs-stage/src/unifyfs-stage-transfer.c

```diff
@@ -0,0 +1,294 @@
+        return ret;
+    }
+
+    ret = md5_checksum(dst, dst_digest);
+    if (ret) {
+        fprintf(stderr, "failed to calculate checksum for %s (%s)\n",
+                        dst, strerror(ret));
+        return ret;
+    }
+
+    if (verbose) {
+        printf("[%d] src: %s, dst: %s\n", rank,
+               checksum_str(md5src, src_digest),
+               checksum_str(md5dst, dst_digest));
+    }
+
+    for (i = 0; i < MD5_DIGEST_LENGTH; i++) {
+        if (src_digest[i] != dst_digest[i]) {
+            fprintf(stderr, "[%d] checksum verification failed: "
+                            "(src=%s, dst=%s)\n", rank,
+                            checksum_str(md5src, src_digest),
+                            checksum_str(md5dst, dst_digest));
+            ret = EIO;
+        }
+    }
+
+    return ret;
+}
+
+static inline char* str_rtrim(char* str)
+{
+    if (str) {
+        char* pos = &str[strlen(str) - 1];
+
+        while (pos >= str && isspace(*pos)) {
+            *pos = '\0';
+            pos--;
+        }
+    }
+
+    return str;
+}
+
+/*
+ * TODO: this currently assumes that line are well-formatted, i.e., no
+ * unnecessary spaces. we would need to add logic to trim the line.
+ */
+static inline int parse_line(char* line, char** src, char** dst)
+{
+    char* pos = strchr(line, ',');
+
+    if (!pos) {
+        return EINVAL;
+    }
+
+    str_rtrim(line);
+
+    pos[0] = '\0';
+
+    *src = line;
+    *dst = &pos[1];
+
+    return 0;
+}
+
+int unifyfs_stage_transfer(unifyfs_stage_t* ctx)
+{
+    int ret = 0;
+    int count = 0;
+    FILE* fp = NULL;
+    char* src = NULL;
+    char* dst = NULL;
+    char linebuf[LINE_MAX] = { 0, };
+    struct stat sb = { 0, };
+
+    if (!ctx) {
+        return EINVAL;
+    }
+
+    fp = fopen(ctx->manifest_file, "r");
+    if (!fp) {
+        fprintf(stderr, "failed to open file %s: %s\n",
+                        ctx->manifest_file, strerror(errno));
+        ret = errno;
+        goto out;
+    }
+
+    while (NULL != fgets(linebuf, LINE_MAX-1, fp)) {
+        ret = parse_line(linebuf, &src, &dst);
+        if (ret) {
+            fprintf(stderr, "failed to parse %s (%s)\n",
+                            linebuf, strerror(ret));
+            goto out;
+        }
+
+        if (ctx->mode == UNIFYFS_STAGE_SERIAL) {
+            if (count % total_ranks == rank) {
+                if (verbose) {
+                    fprintf(stdout, "[%d] serial transfer: src=%s, dst=%s\n",
+                                    rank, src, dst);
+                }
+
+                ret = unifyfs_transfer_file_serial(src, dst);
+                if (ret) {
+                    goto out;
+                }
+
+                ret = stat(dst, &sb);
+                if (ret < 0) {
+                    fprintf(stderr, "stat on %s failed (err=%d, %s)\n",
+                                    dst, errno, strerror(errno));
+                    ret = errno;
+                    goto out;
+                }
```

Comment:
This `stat()` isn't necessary.  If the file isn't there, it's going to be caught in `verify_checksum()`.

Reply:
@tonyhutter Correct for regular file systems. But for unifyfs it seems necessary to have the file correctly laminated. Someone actually needs to confirm this.

Reply:
I'm a little confused.  Are you saying the `stat()` helps to laminate the file after it's transferred?

Reply:
Honestly, I am a bit confused as well. I've had some problems with laminating files in my test setup. It seems like we need to do this for N-1 files, according to the write example:

https://github.com/LLNL/UnifyFS/blob/0c9eb9ab10a9008ddb1789efebe62e10ab7a6eac/examples/src/testutil_rdwr.h#L236

```

    if (cfg->io_pattern == IO_PATTERN_N1) {
        test_barrier(cfg);
        if (cfg->rank != 0) {
            /* call stat() to update global metadata */
            struct stat st;
            int stat_rc = stat(filepath, &st);
            if (-1 == stat_rc) {
                /* lamination failed */
                test_print(cfg, "stat() update during lamination failed");
                rc = -1;
            }
        }

```

I haven't had enough time for figuring out this.

Reply:
You shouldn't need to do anything special.  Your current callpath will laminate the file:
```
unifyfs_transfer_file_serial()
    unifyfs_transfer_file()
        chmod(0222)
```
If for some reason the files aren't laminated, you could try calling `sync()` after they're all transferred.

Reply:
I looked the current client code, and it looks like the `unifyfs_fid_sync` function is the one that sends all cached extent information down to the server daemon.

https://github.com/LLNL/UnifyFS/blob/0c9eb9ab10a9008ddb1789efebe62e10ab7a6eac/client/src/unifyfs.c#L1700

This function is called from multiple system calls including `stat(2)`. Technically, `stat(2)` here doesn't look necessary, because `unifyfs_fid_sync` should be invoked by `close(2)` and `fsync(2)` as well.

I will remove the `stat(2)` here and verify if it still works. Also, one thing we can do here is we can see if the dest file size (by stat) is identical to the source file size. If the size is different, we don't have to call the expensive checksum calculation which will obviously fail.

Reply:
And, the lamination process still confuses me. Basically, the file should be laminated before accessing (reading data) from all nodes. But simply closing a file will not laminate the file. Also, laminating a file won't push the unwritten metadata to the server. Further, you can laminate a currently opened file with `fchmod(2)`. Besides the fact we laminate a file with chmod (instead of having a dedicated API), I don't know if all those things are intuitive enough.